### PR TITLE
Refactor node builders [ECR-4298]

### PR DESCRIPTION
**Pull request by slowli**
## Overview

This PR refactors node builders in the testkit and the CLI crate, extracting reusable parts into a separate extendable interface.

---

See: https://jira.bf.local/browse/ECR-4298

### CODE REVIEW

File: runtimes/rust/src/spec.rs [hunk 1]
```diff
@@ -0,0 +1,247 @@
...
+///
+/// This is a sealed trait: it is not meant to be implemented by the external data types.
+pub trait Deploy: Sealed {
+    /// Modifies the genesis config and Rust runtime builders to deploy self.
+    fn deploy(self, genesis: &mut GenesisConfigBuilder, rt: &mut RustRuntimeBuilder);
+}
+
+/// Marker type for artifact deployment without migration support.
+#[derive(Debug)]
+pub struct Simple(());
```
Comment (alekseysidorov) [inline runtimes/rust/src/spec.rs]: Why not just `Service`?

File: runtimes/rust/src/spec.rs [hunk 2]
```diff
@@ -0,0 +1,247 @@
...
+
+/// Marker type for artifact deployment with migration support.
+#[derive(Debug)]
+pub struct Migrating(());
+
+/// Deploy specification for a Rust artifact. The spec can include 0 or more instantiated services.
```
Comment (alekseysidorov) [inline runtimes/rust/src/spec.rs]: Imo zero instead of 0 is more readable.

File: runtimes/rust/src/spec.rs (hunk 1 above)
Comment (slowli) [reply runtimes/rust/src/spec.rs]: I think `Simple` / `Migrating` dichotomy provides a better description than `Service` / `Migrating`. (For one, "simple" and "migrating" are both adjectives 🙃)

Comment (alekseysidorov) [reply runtimes/rust/src/spec.rs]: Oh, that is my mistake, I meant that we can just write `struct Simple` instead of `struct Simple(())`.

Comment (slowli) [reply runtimes/rust/src/spec.rs]: I wanted `Simple` / `Migrating` to be marker types (i.e., not instantiable).

**Pull request merged**